File: internal/api/ipfilter.go
package api

import (
	"net/http"

	"blackbird/internal/history"
	"blackbird/internal/ipfilter"
)

// ---- GET /api/ipfilter, POST /api/ipfilter/reload (PAR-5.6) ----

// ipfilterStatusHandler reports the blocklist state for Settings: whether a
// source is configured, the rule count, the last load time, and errors. It
// never fetches or loads; only POST reloads.
func (s *Server) ipfilterStatusHandler(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, s.ipfilterStatus())
}

// ipfilterReloadHandler re-fetches (URL sources) and re-loads the daemon
// table now, then reports the status. Failures are 502s carrying the error
// in the status payload; they are also history events.
func (s *Server) ipfilterReloadHandler(w http.ResponseWriter, r *http.Request) {
	if s.opts.IPFilter == nil {
		writeAPIError(w, http.StatusBadRequest, "ipfilter_disabled", "no blocklist source is configured (network.ipfilter)")
		return
	}
	if s.opts.Store != nil && !s.opts.Store.Get().Network.IPFilter.Enabled() {
		writeAPIError(w, http.StatusBadRequest, "ipfilter_disabled", "no blocklist source is configured (network.ipfilter)")
		return
	}
	if err := s.opts.IPFilter.ApplyNow(r.Context()); err != nil {
		if s.history != nil {
			s.history.Add("", history.Entry{
				Kind: history.KindAction, Actor: actorFromRequest(r, s.auth),
				Action: "ipfilter_reload", Result: "failed", Message: err.Error(),
			})
		}
		writeAPIError(w, http.StatusBadGateway, "ipfilter_failed", err.Error())
		return
	}
	if s.history != nil {
		st := s.opts.IPFilter.Status()
		s.history.Add("", history.Entry{
			Kind: history.KindAction, Actor: actorFromRequest(r, s.auth),
			Action: "ipfilter_reload", Result: "ok",
			Message: itoa(st.Rules) + " rules loaded",
		})
	}
	writeJSON(w, http.StatusOK, s.ipfilterStatus())
}

// ipfilterStatus renders the status shape with the service behind a nil
// guard so tests without the service still get a disabled payload.
func (s *Server) ipfilterStatus() ipfilter.Status {
	if s.opts.IPFilter == nil {
		return ipfilter.Status{}
	}
	return s.opts.IPFilter.Status()
}

File: web/src/store/ipfilter.ts
// IP-filter store: the blocklist state (PAR-5.6) for Settings > Connection.
// Refreshes are silent GETs; only reloadNow POSTs, always from an explicit
// user action.

import { createSignal } from "solid-js";
import { showToast } from "./ui";

export interface IPFilterStatus {
  enabled: boolean;
  source: string; // file | url | ""
  path?: string;
  rules: number;
  lastLoad?: string; // RFC3339
  lastError?: string;
}

const [status, setStatus] = createSignal<IPFilterStatus | null>(null);
const [reloading, setReloading] = createSignal(false);

export { reloading, status };

export async function refreshIPFilter(): Promise<void> {
  try {
    const response = await fetch("/api/v1/ipfilter", {
      headers: { Accept: "application/json" },
    });
    if (!response.ok) return;
    const body = (await response.json()) as IPFilterStatus;
    setStatus(body);
  } catch {
    /* leave the previous status intact */
  }
}

/** Re-fetches (URL sources) and re-loads the daemon table now. */
export async function reloadIPFilter(): Promise<void> {
  setReloading(true);
  try {
    const response = await fetch("/api/v1/ipfilter/reload", { method: "POST" });
    const body = (await response.json().catch(() => ({}))) as IPFilterStatus & {
      error?: { message?: string };
      message?: string;
    };
    if (!response.ok) {
      await refreshIPFilter();
      throw new Error(body.error?.message || body.message || "Blocklist reload failed");
    }
    setStatus(body);
    showToast(`Blocklist loaded: ${body.rules} rules.`);
  } catch (error) {
    showToast(error instanceof Error ? error.message : "Blocklist reload failed.");
  } finally {
    setReloading(false);
  }
}
